### PR TITLE
Revised matchings interfaces

### PR DESCRIPTION
**Pull request by d-michail**
Followup of pull-request #281 which adjusts the matching interfaces.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/MaximumWeightBipartiteMatching.java [hunk 1]
```diff
@@ -98,6 +98,15 @@ public double getMatchingWeight()
...
         return weight;
     }
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Matching<E> getMatching(Graph<V, E> graph)
+    {
+        throw new UnsupportedOperationException();
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/MaximumWeightBipartiteMatching.java]: hm, why is this unsupported?

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/BipartiteMatchingAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * Compute a matching for a given graph.
+     * 
+     * @param graph the input graph
+     * @param partition1 the first partition of the vertices
+     * @param partition2 the second partition of the vertices
+     * @return a matching
+     */
+    Matching<E> getMatching(
+        Graph<V, E> graph, Set<? extends V> partition1, Set<? extends V> partition2);
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/BipartiteMatchingAlgorithm.java]: Just for my understanding, why not:
Graph<V, E> graph, Set< V> partition1, Set<V> partition2);
After all, the algorithm is invoked on a graph <V,E>?

**Comment by jkinable** [comment]
It seems that you have a preference for the following structure:
class MyAlgorithm{
 public MyAlgorithm(){ }
 public Result calculateSomething(Graph<V,E> graph){..}
}

You explicitly provide the graph as a parameter to the method that calculates the result. Many of the older classes in jgrapht however adhere to the structure where the graph is passed to the constructor, i.e.:
class MyAlgorithm{
 public MyAlgorithm(Graph<V,E> graph){ }
 public Result calculateSomething(){..}
}

I personally prefer the 2nd approach. I find this more intuitively, since the algorithm can maintain a state. For instance in the floydWarshall class, you only want to compute the distance matrix once, and have a method getShortestPath(V start, V end) which can be invoked repeatedly without recomputing the distance matrix. having a method getShortestPath(Graph graph, V start, V end) would make this much less apparent. I don't think it's a bad convention that if a user wants to run the algorithm on an entirely new graph, he should just make a new instance of the algorithm.
What do you think?

**Comment by d-michail** [comment]
No strong arguments here, I will change it to the second version with the graph instance in the constructor.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/BipartiteMatchingAlgorithm.java (hunk 2 above)
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/BipartiteMatchingAlgorithm.java]: The `KuhnMunkresMinimalWeightBipartitePerfectMatching` class already accepts the partitions with `<? extends V>` so I tried to unify.

**Comment by d-michail** [comment]
I updated the algorithms and  interfaces to accept the graph instance in the constructor. 

However, now the interface change has to be done in 2 steps. I named the new method `computeMatching` and will rename it to `getMatching` after the next release.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MatchingAlgorithm.java [hunk 3]
```diff
@@ -17,24 +17,63 @@
...
+     * @return a matching
+     */
+    Matching<E> computeMatching();
+
+    /**
+     * A graph matching.
+     *
+     * @param <E> the graph edge type
+     */
+    interface Matching<E>
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MatchingAlgorithm.java]: Consider adding a default implementation as was done with MaximumFlow:
MatchingImpl extends Matching.
I think @jsichi started this convention for the flow implementations.
If you add a default implementation, please add a toString() method as well.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MatchingAlgorithm.java [hunk 4]
```diff
@@ -17,24 +17,63 @@
...
  * @param <V> the graph vertex type
  * @param <E> the graph edge type
- * 
  */
 public interface MatchingAlgorithm<V, E>
 {
     /**
+     * Default tolerance used by algorithms comparing floating point values.
+     */
+    double DEFAULT_EPSILON = 1e-9;
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MatchingAlgorithm.java]: Hm, good idea to add the DEFAULT_EPSILON here. We should probably do this for the flow/cut implementations as well.

File: jgrapht-core/src/main/java/org/jgrapht/alg/KuhnMunkresMinimalWeightBipartitePerfectMatching.java [hunk 5]
```diff
@@ -122,6 +122,34 @@ public double getMatchingWeight()
...
+    @Override
+    public Matching<E> computeMatching()
+    {
+        Set<E> m = getMatching();
+        double w = 0d;
+        for (E e : m) {
+            w += graph.getEdgeWeight(e);
+        }
+        final double weight = w;
+        return new Matching<E>()
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/KuhnMunkresMinimalWeightBipartitePerfectMatching.java]: hm, why don't you return a new DefaultMatching(m, weight); here?

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<E> getEdges()
+    {
+        return edges;
+    }
+
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java]: Could you add a basic toString implementation as well? Makes it easier to debug applications.

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import org.jgrapht.alg.interfaces.MatchingAlgorithm.Matching;
+
+/**
+ * A default implementation of the matching interface.
+ * 
+ * @param <E> the graph edge type
+ *
+ * @author Dimitrios Michail
+ */
+class DefaultMatching<E>
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java]: In the MaximumFlowAlgorithm interface, @jsichi  and Alexey added both the MaximumFlow interface as well as a MaximumFlowImpl which implements the MaximumFlow interface. Here there are some consistency issues:

1. Naming: DefaultMatching vs MatchingImpl (same as MaximumFlowImpl)
2. Here the DefaultMatching is a dedicated class, whereas in the Flow case, the implementation is located in the MaximumFlowAlgorithm interface.

I would suggest to keep everything as consistent as possible. Let's choose a standard name, e.g DefaultMatching or MatchingImpl and enforce the same for the other interfaces. Similarly, should the default implementation reside in the interface, or in a separate class? @jsichi preferences?

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/EdmondsBlossomShrinking.java [hunk 8]
```diff
@@ -36,43 +36,44 @@
...
     }
 
     /**
      * Runs the algorithm on the input graph and returns the match edge set.
      *
-     * @return set of Edges
+     * @return set of edges
      */
-    private Set<E> findMatch()
+    private Set<E> findMatch(UndirectedGraph<V, E> graph)
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/EdmondsBlossomShrinking.java]: graph parameter can be removed as it already is a member variable. Same for findPath method.

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/EdmondsBlossomShrinking.java [hunk 9]
```diff
@@ -36,43 +36,44 @@
 public class EdmondsBlossomShrinking<V, E>
     implements MatchingAlgorithm<V, E>
 {
-
     private UndirectedGraph<V, E> graph;
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/EdmondsBlossomShrinking.java]: Could you make graph final?

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/HopcroftKarpBipartiteMatching.java [hunk 10]
```diff
@@ -37,40 +38,53 @@
...
  * @author Joris Kinable
  */
-
 public class HopcroftKarpBipartiteMatching<V, E>
     implements MatchingAlgorithm<V, E>
 {
-    private final UndirectedGraph<V, E> graph;
-    private final Set<V> partition1; // Partitions of bipartite graph
-    private final Set<V> partition2;
+    private UndirectedGraph<V, E> graph;
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/HopcroftKarpBipartiteMatching.java]: Could you make graph final?

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/KuhnMunkresMinimalWeightBipartitePerfectMatching.java [hunk 11]
```diff
@@ -43,81 +43,85 @@
...
     public KuhnMunkresMinimalWeightBipartitePerfectMatching(
-        final WeightedGraph<V, E> G, List<? extends V> S, List<? extends V> T)
+        Graph<V, E> graph, Set<? extends V> partition1, Set<? extends V> partition2)
+    {
+        if (graph == null) {
+            throw new IllegalArgumentException("Input graph cannot be null");
+        }
+        this.graph = graph;
+        if (partition1 == null) {
+            throw new IllegalArgumentException("Invalid partition provided");
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/KuhnMunkresMinimalWeightBipartitePerfectMatching.java]: Could you change the message to: "Partition 1 cannot be null"
Same for partition 2.

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/PathGrowingWeightedMatching.java [hunk 12]
```diff
@@ -178,10 +164,10 @@ public double getMatchingWeight()
     }
 
     // the algorithm (no heuristics)
-    private Pair<Double, Set<E>> run(Graph<V, E> graph)
+    private Pair<Double, Set<E>> run()
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/PathGrowingWeightedMatching.java]: Can we eliminate the usage of Pair here and immediately return a DefaultMatching instead?

**Comment by jkinable** [comment]
Alright, I left a few comments. Definitely a good improvement in terms of streamlining the library :)

**Comment by jkinable** [comment]
Could you add a comment to the MatchingAlgorithm interface that "computeMatching" should be renamed to getMatching in the next release? I'm sure I'll forget otherwise...

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java (hunk 7 above)
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java]: I agree, let's follow the pattern from MaximumFlow.

Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java]: I prefer to keep the default implementation in a separate class in order to keep it package-private. This way the user is forced to use the interface and not the Data-Transfer-Object. Moreover, this way we can use only one DTO for other interfaces which are also weighted edge sets. Similarly for vertex sets such as cliques, vertex-covers, etc.

Another argument for this is that it keeps the public footprint smaller and it is easier to keep compatibility between versions of the library.

I would vote to go the other way and move the MaximumFlowImpl in the flow packages as package-private. Of course, if both you and @jsichi prefer to use the flow pattern, we can do that.

File: jgrapht-core/src/main/java/org/jgrapht/alg/KuhnMunkresMinimalWeightBipartitePerfectMatching.java (hunk 5 above)
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/KuhnMunkresMinimalWeightBipartitePerfectMatching.java]: This is the deprecated implementation, the new class uses directly the `DefaultMatching` class.

**Comment by d-michail** [comment]
Thanks for the review. I did all the changes apart from moving the matching DTO into the interface. Tell me if you think this is also necessary.

File: jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java (hunk 7 above)
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/matching/DefaultMatching.java]: There are two kinds of users: those who use existing algorithms, and those who develop customized algorithms of their own (not intended for inclusion in JGraphT).  Keeping the DTO public makes it reusable by the second kind of user, which I think is a good thing.

(For the first kind of user, they would have to explicitly downcast the algorithm result to get at the DTO, and I don't see why they would even want to do this?  So I think the public footprint benefit for the second kind of user outweighs the downside.)

Regarding "we can use only one DTO for other interfaces which are also weighted edge sets", that makes sense...in that case, if we want to use a top-level class, then the class name should be unrelated to matching.

**Comment by jkinable** [comment]
I agree with John: from a user perspective, it is much more convenient to
have direct access to the DTO, so I would prefer to make this object public.
Coming up with a generic, descriptive name may be pretty hard. Something
like WeightedEdgeSet would be a rather poor name as it does not reflect the
fact that it is the result of an algorithm: a subset of edges with an
associated 'score'. I would probably prefer having multiple (duplicate)
implementations MatchingImpl, CoverImpl just for the added clarity.

Btw, you renamed DefaultMatching to DefaultMatchingImpl. Should we rename
MaximumFlowImpl to DefaultMaximumFlowImpl or DefaultFlowImpl for
consistency?

**Comment by d-michail** [comment]
ok, I moved the matching implementation into the matching interface. I also named it `MatchingImpl` to be consistent with the flows. 

All comments have been addressed.

**Comment by jkinable** [comment]
Looks very nice. Thanks!

**Pull request merged**